identity-credential-directive.js

```diff
@@ -0,0 +1,247 @@
+
+    scope.claimsForCredential = function(credential) {
+      // TODO: This only pulls in claims at the top level, and does not recover nested claims
+      var claims = [];
+      for(var key in credential.claim) {
+        // TODO: Need better filtering here (probably though some kind of grammar
+        // that maps claim types to a readable output), right now we're just taking
+        // the first level claims w/o any of their nested properties
+        if(key === 'id') {
+          continue;
+        }
+        if(key === 'image') {
+          continue;
+        }
+        claims.push(key)
+      }
+
+      var props = [];
+      traverse(credential.claim);
+      return props;
+
+
+      function traverse(o) {
+        for(var i in o) {
+          if(typeof(o[i]) === 'object') {
+              props.push(i);
+              traverse(o[i]);
+          } else if(!(i === 'id' || i === 'type' || i === '@value')) {
+              props.push(i);
+          }
+        }
+      }
+    };
+
+    scope.replacementCredentials = function(credential) {
+      var substituteCredentials = [];
+
+      // TODO: This doesn't really handle queries that take in a specific value request, because it only looks
+      // if credentials have matching keys (but shouldn't the consumer be verifying the returned values anyway?)
+
+      // Get all of the passed in credential's claims
+      var claims = scope.claimsForCredential(credential);
+      // Filter the passed in claims that match with the properties requested in the query, minus
+      // any claims that are already fulfilled by the currently selected credentials
+      var requestedClaims = claims.filter(function(claim) {
+        for(var key in scope.selectedCredentials) {
+          var selectedCredential = scope.selectedCredentials[key];
+          if(selectedCredential === credential) {
+            continue;
+          }
+          if(jsonld.hasProperty(selectedCredential.claim, claim)) {
+            // Property already fulfilled
+            return false;
+          }
+        }
+        return jsonld.hasProperty(scope.query, claim);
+      });
+
+      // Filter through all of the user's credentials, returning those that fulfill the requested claims
+      for(var key in scope.allCredentials) {
+        var substituteCredential = scope.allCredentials[key];
+        var fulfillable = true;
+        for(key in requestedClaims) {
+          var requestedClaim = requestedClaims[key];
+          if(!(jsonld.hasProperty(substituteCredential.claim, requestedClaim) &&
+              substituteCredentials.indexOf(substituteCredential) === -1 &&
+              substituteCredential !== credential)) {
+            fulfillable = false;
+          }
+        }
+        if(fulfillable) {
+          substituteCredentials.push(substituteCredential);
+        }
+      }
+      return substituteCredentials;
+    }
+
+
+    scope.credentialsReplaceableByCredential = function(credential) {
```

Comment:
Use verb/interrogative for name ... maybe `getReplacementCandidates`? Or something to that effect.